src/server/ns_turn_ratelimit.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2024 Wire Swiss GmbH
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the project nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "ns_turn_maps.h"
+#include "ns_turn_ioalib.h"
+#include "ns_turn_ratelimit.h"
+
+/////////////////// rate limit //////////////////////////
+
+ur_addr_map *rate_limit_map = NULL;
+TURN_MUTEX_DECLARE(rate_limit_main_mutex);
+
+void ratelimit_add_node(ioa_addr *address) {
+  // copy address
+  ratelimit_entry *rateLimitEntry = (ratelimit_entry *)malloc(sizeof(ratelimit_entry));
+  TURN_MUTEX_INIT(&(rateLimitEntry->mutex));
+  rateLimitEntry->request_count = 1;
+  rateLimitEntry->last_request_time = time(NULL);
+
+  ur_addr_map_put_no_port(rate_limit_map, address, (ur_addr_map_value_type)rateLimitEntry);
+}
+
+int ratelimit_delete_expired(ur_map_value_type value) {
+  time_t current_time = time(NULL);
+  ratelimit_entry *rateLimitEntry = (ratelimit_entry*)(void*)(ur_map_value_type)value;
+  if (rateLimitEntry->last_request_time + RATE_LIMIT_MAX_REQUESTS_PER_WINDOW < current_time)
```

Comment:
Is this calculation correct? Shouldn't it be

```suggestion
  if (rateLimitEntry->last_request_time + RATE_LIMIT_WINDOW_SECS < current_time)
```
:thinking:

Reply:
Correct!